libflashrom.c

```diff
@@ -103,429 +106,455 @@ int print(const enum flashrom_log_level level, const char *const fmt, ...)
 }
 
-/* TODO: flashrom_programmer_capabilities()? */
-
-/** @} */ /* end flashrom-prog */
-
-
-
-/**
- * @defgroup flashrom-flash Flash chips
- * @{
- */
-
 /**
- * @brief Probe for a flash chip.
- *
- * Probes for a flash chip and returns a flash context, that can be used
- * later with flash chip and @ref flashrom-ops "image operations", if
- * exactly one matching chip is found.
- *
- * @param[out] flashctx Points to a pointer of type struct flashrom_flashctx
- *                      that will be set if exactly one chip is found. *flashctx
- *                      has to be freed by the caller with @ref flashrom_flash_release.
- * @param[in] flashprog The flash programmer used to access the chip.
- * @param[in] chip_name Name of a chip to probe for, or NULL to probe for
- *                      all known chips.
- * @return 0 on success,
- *         3 if multiple chips were found,
- *         2 if no chip was found,
- *         or 1 on any other error.
+ * @brief Returns list of supported programmers
+  * @return List of supported programmers
  */
-int flashrom_flash_probe(struct flashrom_flashctx **const flashctx,
-			 const struct flashrom_programmer *const flashprog,
-			 const char *const chip_name)
+const char **flashrom_supported_programmers(void)
 {
-	int i, ret = 2;
-	struct flashrom_flashctx second_flashctx = { 0, };
+	enum programmer p = 0;
+  const char **supported_programmers = NULL;
+  supported_programmers = malloc((PROGRAMMER_INVALID + 1) * sizeof(char*));
 
-	chip_to_probe = chip_name; /* chip_to_probe is global in flashrom.c */
-
-	*flashctx = malloc(sizeof(**flashctx));
-	if (!*flashctx)
-		return 1;
-	memset(*flashctx, 0, sizeof(**flashctx));
-
-	for (i = 0; i < registered_master_count; ++i) {
-		int flash_idx = -1;
-		if (!ret || (flash_idx = probe_flash(&registered_masters[i], 0, *flashctx, 0)) != -1) {
-			ret = 0;
-			/* We found one chip, now check that there is no second match. */
-			if (probe_flash(&registered_masters[i], flash_idx + 1, &second_flashctx, 0) != -1) {
-				ret = 3;
-				break;
-			}
+	if (supported_programmers != NULL) {
+		for (; p < PROGRAMMER_INVALID; ++p){
+		  supported_programmers[p] = programmer_table[p].name;
 		}
+	} else {
+		msg_gerr("Memory allocation error!");
 	}
-	if (ret) {
-		free(*flashctx);
-		*flashctx = NULL;
-	}
-	return ret;
+
+	return supported_programmers;
 }
 
 /**
- * @brief Returns the size of the specified flash chip in bytes.
- *
- * @param flashctx The queried flash context.
- * @return Size of flash chip in bytes.
+ * @brief Returns list of supported flash ROM chips
+ * @return List of supported flash ROM chips
  */
-size_t flashrom_flash_getsize(const struct flashrom_flashctx *const flashctx)
+flashrom_flashchip_info *flashrom_supported_flash_chips(void)
 {
-	return flashctx->chip->total_size * 1024;
+	int i = 0;
+  flashrom_flashchip_info *supported_flashchips = NULL;
+	supported_flashchips = malloc(flashchips_size * sizeof(supported_flashchips));
```

Comment:
@artur-rs this looks incorrect, maye you meant:

```suggestion
	supported_flashchips = malloc(flashchips_size * sizeof(flashrom_flashchip_info));
```